### PR TITLE
feat(ITIL Template): read-only fields

### PR DESCRIPTION
**Pull request by Rom1-B**
Allows managing read-only fields via templates

image

image
_It is not obvious on the screenshot, but the opening date field is disabled_

This PR does not handle special cases such as actors, as not requested by the customer

| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | !25392

### CODE REVIEW

**Review by cconard96 — APPROVED** [review]

**Comment by trasher** [comment]
Tests are failing

File: src/ITILTemplateReadonlyField.php [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+
+        $readonlyfields = [];
+        $used         = [];
+        foreach ($iterator as $data) {
+            $readonlyfields[$data['id']] = $data;
+            $used[$data['num']]        = $data['num'];
+        }
+
+        if ($canedit) {
+            echo "<div class='firstbloc'>";
```
Comment (trasher) [inline src/ITILTemplateReadonlyField.php:199]: We should use twig templates for new developments; unless there is a reason not to.

**Review by cedric-anne — CHANGES_REQUESTED** [review]
1. A control logic should probably be added somewhere to ensure that read-only fields are not updatable if someone forces some values when sending form. This should be done only when update is call from UI, as these read-only fields should probably remains updatable by any backend logic. By the way, same kind of control should probably be done also for hidden fields.

2. `disabled` inputs will not be sent in POST request and threfore will not be pssed to rules engine. I do not know if it will be considered as a problem.

**Comment by trasher** [comment]
> 1. A control logic should probably be added somewhere to ensure that read-only fields are not updatable if someone forces some values when sending form. This should be done only when update is call from UI, as these read-only fields should probably remains updatable by any backend logic. By the way, same kind of control should probably be done also for hidden fields.
> 
>     2. `disabled` inputs will not be sent in POST request and threfore will not be pssed to rules engine. I do not know if it will be considered as a problem.

Are those fixed? I did not pay attention to this comment for the first time, but we should really not send hidden/read-only data from the form (or at least, do not use it, to prevent any user able to manipulate the HTML page to send data).

Also, I guess a few tests may be added.

**Comment by trasher** [comment]
And ho, for the rules engine, I think we must send the values, or the rules may not be applied correctly... At creation, it must be the ones defined in the template; and the ones already in database in update case (template may have changed since item creation)

**Comment by Rom1-B** [comment]
If, for example, the user triggers a rule that updates a hidden or read-only field. Should this action be blocked?

IMHO, the template is more about facilitating input than controlling it?
@orthagh @cedric-anne

**Review by cedric-anne — APPROVED** [review]

**Comment by cedric-anne** [comment]
> If, for example, the user triggers a rule that updates a hidden or read-only field. Should this action be blocked?
> 
> IMHO, the template is more about facilitating input than controlling it? @orthagh @cedric-anne

As discussed in our weekly meeting, ITIL templates are made only to control form displaying.

**Review by trasher — APPROVED** [review]

**Pull request merged**